sound/soc/sof/pm.c

```diff
@@ -143,6 +143,8 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 		struct sof_ipc_reply reply;
 		struct sof_ipc_dai_config *config = dai->dai_config;
 
+		config->dai_index = dai->comp_dai.dai_index;
```

Comment:
@zhuyingjiang if you look at sof_link_hda_load(), we send 2 ipcs for both cases:
 	if (link->dpcm_capture) and 	if (link->dpcm_playback) during topology load.

But at resume we only send 1 ipc. thats likely whats missing at resume.

Reply:
@ranj063 No, at resume we still send 2, but there are two tx messages with the same dai_index, so caused error.

Reply:
@plbossart 
1. The fix do the same in sof_restore_pipelines as in sof_link_hda_process, which was missing in the previous sof_restore_pipelines code. And I have tested on WHL RVP this can fix issue #442. 
    The resume use the lists to restore the pipelines, but in sof_link_hda_process, the tx message sent is not the same as stored in the list, the difference is the config->dai_index = dai->comp_dai.dai_index; So when do resume, we need assign it again.
2. I agree with you we need go deep and find the root cause, maybe this is a issue with topology.c.
As I add a log print before "config->dai_index = dai->comp_dai.dai_index" in sof_link_hda_process, the result is:
[    3.709139] sof-audio sof-audio: ======config->dai_index: **00000000**, dai->comp_dai.dai_index: **00000000**=====
[    3.709192] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709350] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709353] sof-audio sof-audio: ======config->dai_index: **00000000**, dai->comp_dai.dai_index: **00000001**=====
[    3.709407] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709562] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709566] sof-audio sof-audio: warn: unexpected DAI config count 0!
[    3.709569] sof-audio-pci 0000:00:1f.3: link dma channel 1 for playback
[    3.709571] sof-audio-pci 0000:00:1f.3: link dma channel 1 for capture
[    3.709572] sof-audio sof-audio: ======config->dai_index: **00000001**, dai->comp_dai.dai_index: **00000002**=====
[    3.709626] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709776] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709780] sof-audio sof-audio: ======config->dai_index: **00000002**, dai->comp_dai.dai_index: **00000003**=====
[    3.709834] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.709994] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.710001] sof-audio sof-audio: warn: unexpected DAI config count 0!
[    3.710005] sof-audio-pci 0000:00:1f.3: link dma channel 2 for playback
[    3.710007] sof-audio sof-audio: ======config->dai_index: **00000004**, dai->comp_dai.dai_index: **00000004**=====
[    3.710039] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.710193] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.710198] sof-audio sof-audio: warn: unexpected DAI config count 0!
[    3.710201] sof-audio-pci 0000:00:1f.3: link dma channel 3 for playback
[    3.710202] sof-audio sof-audio: ======config->dai_index: **00000005**, dai->comp_dai.dai_index: **00000005**=====
[    3.710256] sof-audio sof-audio: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.710411] sof-audio sof-audio: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.710416] sof-audio sof-audio: warn: unexpected DAI config count 0!
[    3.710419] sof-audio-pci 0000:00:1f.3: link dma channel 4 for playback
[    3.710420] sof-audio sof-audio: ======config->dai_index: **00000006**, dai->comp_dai.dai_index: **00000006**=====
It seems some dai_index are the same with comp_dai.dai_index, but others are not. The difference make this issue. I will go deep to see if it is a bug of topology or it is intended to do so.
3. Maybe it is because on other platform the config->dai_index are the same with comp_dai.dai_index, so this will not cause error, and this depend on how the tplg file define the pipeline.  I will check on other platform to resolve this.

Reply:
@zhuyingjiang if we're sengin 2 ipcs then the problem is probably that we're overwriting the config->dai_index with the second one and it fails during resume. 

So maybe we should save both playback dai_index and capture dai_index separately for hda dai's or we re-assign the dai_index again upon resume.

Reply:
@ranj063 The problem is that we sometimes assign same dai_index to playback and capture, and some time different. (The first two in the previous log using the same dai_index for playback and capture, and the 3 and 4 using different.) Then in topology.c there is a tricky re-assign the dai_index as comp_dai.dai_index and send as the IPC message, not stored in the list. In sof_restore_pipelines there is no such assign then there is a problem.
Is this a historical problem? Or we can force  same playback/capture channel to use different dai_index.

Reply:
@zhuyingjiang I don't what you mean by historical problem. We didn't have hda.capture in glk. So this is new for whl. And yes the said index for playback and capture dais are different for hda dai links. That's why we need resolve the dai index before sending the ipc.

Reply:
@keyonjie @lgirdwood Could you share your comments?

Reply:
@RanderWang can you please help yingjiang with this. You are the one who has most recently worked on the hda dai config

Reply:
> I don't what you mean by historical problem.

@ranj063 I mean this is happened in topology.c, obviously we have in-conformity for some place we have same dai_index playback/capture for same channel, while other place we have different dai_index for playback/capture for same channel. 
If fix with root cause, we need to force playback/capture with same channel to use different dai_index.

Reply:
@zhuyingjiang I dont think we can force the same channel for playback and capture

this is the comment in topology.c that makes the distinction for HDA dai's:
/* for hda link, playback and capture are supported by different
	 * dai in FW. Here send dai config according to capability of dai.
	 */

We handle this during topology parsing. We need to find a way to do the same thing upon resume as well.

Reply:
@ranj063 @plbossart 
If "find a way to do the same thing upon resume as well", this PR is just for that. 
It is simply missing the assign "config->dai_index = dai->comp_dai.dai_index;" in the pm code.

Reply:
> In sof_link_hda_process() the code does this:
> 
> config->dai_index = dai->comp_dai.dai_index;
> 
> [linux/sound/soc/sof/topology.c](https://github.com/thesofproject/linux/blob/b8db053ebce3636f5d3eca2c25b544cd15ee14ef/sound/soc/sof/topology.c#L2151)
> 
> Line 2151 in [b8db053](/thesofproject/linux/commit/b8db053ebce3636f5d3eca2c25b544cd15ee14ef)
>  config->dai_index = dai->comp_dai.dai_index; 
> 
> Why would one need to do this again on resume?
> And why does this specifically occur on WHL? this looks suspicious.
> 
> BTW we should really clean this up, we have two dai_index fields in different structures, it's not really good to track code:
> 
> ```c
> /* DAI */
> static const struct sof_topology_token dai_tokens[] = {
> 	{SOF_TKN_DAI_DMAC_CONFIG, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
> 		offsetof(struct sof_ipc_comp_dai, dmac_config), 0},
> 	{SOF_TKN_DAI_TYPE, SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_dai_type,
> 		offsetof(struct sof_ipc_comp_dai, type), 0},
> 	{SOF_TKN_DAI_INDEX, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
> 		offsetof(struct sof_ipc_comp_dai, dai_index), 0},
> 	{SOF_TKN_DAI_DIRECTION, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
> 		offsetof(struct sof_ipc_comp_dai, direction), 0},
> };
> 
> /* BE DAI link */
> static const struct sof_topology_token dai_link_tokens[] = {
> 	{SOF_TKN_DAI_TYPE, SND_SOC_TPLG_TUPLE_TYPE_STRING, get_token_dai_type,
> 		offsetof(struct sof_ipc_dai_config, type), 0},
> 	{SOF_TKN_DAI_INDEX, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
> 		offsetof(struct sof_ipc_dai_config, dai_index), 0},
> };
> ```

yes, it is confusing. One is dai component id, another is dai link id.